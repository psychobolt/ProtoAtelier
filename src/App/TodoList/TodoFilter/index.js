export { default as Filters } from './TodoFilter.filters';
export { default } from './TodoFilter.component';
