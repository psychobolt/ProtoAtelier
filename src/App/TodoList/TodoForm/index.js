export * from './TodoForm.actions';
export { default } from './TodoForm.container';
