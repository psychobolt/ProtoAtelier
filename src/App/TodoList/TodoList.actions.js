export const Actions = {
  TOGGLE_TODO: 'toggleTodo',
};

export const toggleTodo = id => ({
  type: Actions.TOGGLE_TODO,
  todoId: id,
});
