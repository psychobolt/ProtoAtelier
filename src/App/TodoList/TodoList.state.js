// @flow
import type { Todo } from './TodoItem';

export type TodoListState = {
  todos: Todo[]
};

export default [];
