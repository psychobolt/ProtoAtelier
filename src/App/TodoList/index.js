import todoListReducer from './TodoList.reducer';

export { default as TodoFilter, Filters } from './TodoFilter';
export { default as TodoForm } from './TodoForm';
export { default } from './TodoList.container';

export const reducers = {
  todos: todoListReducer,
};
