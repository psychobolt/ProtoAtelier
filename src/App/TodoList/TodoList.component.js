// @flow
import React from 'react';
import styled from 'styled-components';

import TodoItem, { type Todo } from './TodoItem';
import * as styles from './TodoList.style';

type Props = {
  todos: Todo[],
  onTodoClick: (id: number) => void,
  children: any,
  className: string,
}

const List = styled.ul`${styles.ul}`;

const Item = styled(TodoItem)`${styles.li}`;

export default ({ todos, onTodoClick, children, className }: Props) => (
  <x-card class={className}>
    {children}
    <List>
      {todos.length ? todos.map(todo => (
        <li key={todo.id}>
          <Item
            {...todo}
            onClick={() => onTodoClick(todo.id)}
          />
        </li>
      )) : (
        <x-label>
          {'No items'}
        </x-label>
      )}
    </List>
  </x-card>
);
