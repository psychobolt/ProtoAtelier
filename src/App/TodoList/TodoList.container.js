// @flow
import type { Dispatch } from 'redux';
import { connect } from 'react-redux';

import TodoList from './TodoList.component';
import { toggleTodo } from './TodoList.actions';
import type { TodoListState } from './TodoList.state';
import { Filters } from './TodoFilter';

const getVisibleTodos = (todos, filter) => {
  switch (filter) {
    case Filters.COMPLETED:
      return todos.filter(t => t.completed);
    case Filters.ACTIVE:
      return todos.filter(t => !t.completed);
    case Filters.ALL:
    default:
      return todos;
  }
};

type Props = {
  filter: string
};

export const mapStateToProps = (state: TodoListState, ownProps: Props) => ({
  todos: getVisibleTodos(state.todos, ownProps.filter),
});

const mapDispatchToProps = (dispatch: Dispatch<any>) => ({
  onTodoClick: id => {
    dispatch(toggleTodo(id));
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(TodoList);
