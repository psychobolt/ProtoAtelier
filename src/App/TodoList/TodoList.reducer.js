import initialState from './TodoList.state';
import { Actions as TodoListActions } from './TodoList.actions';
import { Actions as TodoFormActions } from './TodoForm';

export default (state = initialState, action) => {
  switch (action.type) {
    case TodoFormActions.ADD_TODO:
      return [
        ...state,
        {
          id: action.todo.id,
          text: action.todo.text,
          completed: false,
        },
      ];
    case TodoListActions.TOGGLE_TODO:
      return state.map(todo =>
        (todo.id === action.todoId ? { ...todo, completed: !todo.completed } : todo));
    default:
      return state;
  }
};
