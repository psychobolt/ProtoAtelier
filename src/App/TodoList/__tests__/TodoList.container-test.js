import React from 'react';
import { Provider } from 'react-redux';
import { shallow, mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import PropTypes from 'prop-types';

import TodoList from '../TodoList.container';
import { toggleTodo } from '../TodoList.actions';

const mockStore = configureMockStore([]);

const completedTodo = {
  id: 0,
  text: 'Completed Item',
  completed: true,
};
const activeTodo = {
  id: 1,
  text: 'Active Item',
  completed: false,
};
const todos = { present: [completedTodo, activeTodo] };

describe('container <TodoList />', () => {
  it('TodoList should render without crashing', () => {
    shallow(<TodoList />);
  });

  it('TodoList should dispatch action on TodoItem click', () => {
    const id = 0;
    const store = mockStore({ todos });
    const wrapper = mount(<Provider store={store}><TodoList /></Provider>, {
      context: { store },
      childContextTypes: { store: PropTypes.object }, // eslint-disable-line react/forbid-prop-types
    });
    wrapper.find('x-checkbox').first().simulate('click');
    expect(store.getActions()).toEqual([toggleTodo(id)]);
  });
});
