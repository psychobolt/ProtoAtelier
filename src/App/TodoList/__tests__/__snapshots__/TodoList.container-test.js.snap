// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`container <TodoList /> TodoList should render correctly 1`] = `
<TodoList
  onTodoClick={[Function]}
  todos={
    Array [
      Object {
        "completed": true,
        "id": 0,
        "text": "Completed Item",
      },
      Object {
        "completed": false,
        "id": 1,
        "text": "Active Item",
      },
    ]
  }
/>
`;
